server/server.go

```diff
@@ -535,6 +570,63 @@ func (s *Server) handleTopicAuth(w http.ResponseWriter, _ *http.Request, _ *visi
 	return s.writeJSON(w, newSuccessResponse())
 }
 
+func (s *Server) handleAPIWebPushConfig(w http.ResponseWriter, _ *http.Request, _ *visitor) error {
+	response := &apiWebPushConfigResponse{
+		PublicKey: s.config.WebPushPublicKey,
+	}
+
+	return s.writeJSON(w, response)
+}
+
+func (s *Server) handleTopicWebPushSubscribe(w http.ResponseWriter, r *http.Request, v *visitor) error {
+	var username string
+	u := v.User()
+	if u != nil {
+		username = u.Name
+	}
+
+	var sub webPushSubscribePayload
+	err := json.NewDecoder(r.Body).Decode(&sub)
+
+	if err != nil || sub.BrowserSubscription.Endpoint == "" || sub.BrowserSubscription.Keys.P256dh == "" || sub.BrowserSubscription.Keys.Auth == "" {
+		return errHTTPBadRequestWebPushSubscriptionInvalid
+	}
+
+	topic, err := fromContext[*topic](r, contextTopic)
+	if err != nil {
+		return err
+	}
+
+	err = s.webPushSubscriptionStore.AddSubscription(topic.ID, username, sub)
+	if err != nil {
+		return err
+	}
+
+	return s.writeJSON(w, newSuccessResponse())
+}
```

Comment:
I did already wrap this with `limitRequestsWithTopic`, but I haven't looked at the details of how that works. Is that not enough here?

I also wonder if we should limit the endpoints to known browser endpoints as an additional measure.

Reply:
> I did already wrap this with limitRequestsWithTopic, but I haven't looked at the details of how that works. Is that not enough here?

`limitRequestsWithTopic` limits the requests in total, yes. On ntfy.sh this is 250 a day, but on a default instance it's ~17k per day. So I think something needs to be limited here.  *The rule of the Internet is: What can be abused, will be abused.*

I don't have a great answer as to how the limit should look like. The `visitor` already has like ~10 limiters for various actions, but I am always hesitant to add more, because every limiter we adds more RAM usage...

The only sensible option I can think of is this: Add `webPushSubscriptionLimiter *util.RateLimiter` with a fixed rate (not configurable, e.g. 10 per 3 hours or something like that), and only initialize the limiter if s.WebPushEnabled is true

> I also wonder if we should limit the endpoints to known browser endpoints as an additional measure (edit: https://github.com/pushpad/known-push-services/blob/master/whitelist)

This is to limit outgoing requests, right? That seems like the wrong thing to limit. If anything, I'd add 429-handling somehow, but I prefer limiting as early as possible (on the incoming end).

Reply:
> The only sensible option I can think of is this: Add webPushSubscriptionLimiter *util.RateLimiter with a fixed rate (not configurable, e.g. 10 per 3 hours or something like that), and only initialize the limiter if s.WebPushEnabled is true

I guess it's also a balance between how often to update the frontend. I set it to every 5 minutes, but can be much slower - the question is whether it would ever run on the mobile use case of looking a notification and closing the app again (except for the startup refresh). So I wonder if I should have it set to check every 5 minutes plus every time the window comes into the foreground (visibilityState), but no-op if the web push was updated was in the last ~x hours.

10 per 3 hours still seems kinda low nevertheless - or is it per topic?

> This is to limit outgoing requests, right? That seems like the wrong thing to limit. If anything, I'd add 429-handling somehow, but I prefer limiting as early as possible (on the incoming end).

Yes, I was thinking of it as an additional check to prevent using a ntfy server to DDoS another endpoint. There doesn't seem to be a good reason to allow non-known browser push services.

Reply:
> I'd add 429-handling somehow

This is already done, any non 2xx status from the web push endpoint will delete the subscription.